File: sdk/python/atb/encrypt.py
"""
ATB bundle encryption helpers.

Wire format:
    [4 bytes magic "ATBE"]
    [1 byte version]
    [16 bytes salt]
    [12 bytes nonce]
    [16 bytes auth tag]
    [N bytes ciphertext]

Quick start::

    from atb import Bundle, encrypt_bundle
    encrypted = encrypt_bundle(Bundle.load("run.atb/bundle.atb"), "password")
    assert encrypted.startswith(b"ATBE")
"""

from __future__ import annotations

import json
import os
from typing import TYPE_CHECKING

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

from atb.canonicalize import canonicalize
from atb.exceptions import ATBError, ATBVerificationError
from atb.hash import GENESIS_HASH

if TYPE_CHECKING:
    from atb.bundle import Bundle


MAGIC = b"ATBE"
LEGACY_VERSION = 0x01
VERSION = 0x02
SALT_SIZE = 16
NONCE_SIZE = 12
TAG_SIZE = 16
KEY_SIZE = 32
LEGACY_PBKDF2_ITERATIONS = 100_000
PBKDF2_ITERATIONS = 600_000
HEADER_SIZE = len(MAGIC) + 1 + SALT_SIZE + NONCE_SIZE + TAG_SIZE


class ATBEncryptionError(ATBError):
    """Raised when encryption fails.

    Args:
        *args: Positional error message arguments passed to ``Exception``.

    Returns:
        None.

    Raises:
        None.
    """


class ATBDecryptionError(ATBError):
    """Raised when decryption fails.

    Args:
        *args: Positional error message arguments passed to ``Exception``.

    Returns:
        None.

    Raises:
        None.
    """


def _derive_key(
    password: str,
    salt: bytes,
    iterations: int,
    *,
    error_cls: type[ATBError],
) -> bytes:
    if not password:
        raise error_cls("password cannot be empty")
    if len(salt) != SALT_SIZE:
        raise error_cls(f"salt must be {SALT_SIZE} bytes")
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=KEY_SIZE,
        salt=salt,
        iterations=iterations,
    )
    return kdf.derive(password.encode("utf-8"))


def _kdf_iterations_for_version(version: int, *, error_cls: type[ATBError]) -> int:
    if version == LEGACY_VERSION:
        return LEGACY_PBKDF2_ITERATIONS
    if version == VERSION:
        return PBKDF2_ITERATIONS
    raise error_cls(f"unsupported version: 0x{version:02x}")


def _encrypt_raw_with_version(
    plaintext: bytes,
    password: str,
    *,
    version: int,
    salt: bytes | None = None,
    nonce: bytes | None = None,
) -> bytes:
    salt_bytes = salt if salt is not None else os.urandom(SALT_SIZE)
    nonce_bytes = nonce if nonce is not None else os.urandom(NONCE_SIZE)
    if len(salt_bytes) != SALT_SIZE:
        raise ATBEncryptionError(f"salt must be {SALT_SIZE} bytes")
    if len(nonce_bytes) != NONCE_SIZE:
        raise ATBEncryptionError(f"nonce must be {NONCE_SIZE} bytes")

    iterations = _kdf_iterations_for_version(version, error_cls=ATBEncryptionError)
    key = _derive_key(
        password,
        salt_bytes,
        iterations,
        error_cls=ATBEncryptionError,
    )
    aesgcm = AESGCM(key)
    sealed = aesgcm.encrypt(nonce_bytes, plaintext, None)
    if len(sealed) < TAG_SIZE:
        raise ATBEncryptionError("invalid AES-GCM output")
    ciphertext = sealed[:-TAG_SIZE]
    tag = sealed[-TAG_SIZE:]
    return MAGIC + bytes([version]) + salt_bytes + nonce_bytes + tag + ciphertext


def encrypt_raw(
    plaintext: bytes,
    password: str,
    *,
    salt: bytes | None = None,
    nonce: bytes | None = None,
) -> bytes:
    """Encrypt plaintext bytes with AES-256-GCM.

    Optional salt/nonce are intended for deterministic tests and golden vectors.

    Args:
        plaintext: Plain bytes to encrypt.
        password: Password used for PBKDF2 key derivation.
        salt: Optional deterministic salt for tests.
        nonce: Optional deterministic nonce for tests.

    Returns:
        ATBE wire bytes.

    Raises:
        ATBEncryptionError: If the password, salt, nonce, or AES-GCM output is
            invalid.
    """
    return _encrypt_raw_with_version(
        plaintext,
        password,
        version=VERSION,
        salt=salt,
        nonce=nonce,
    )


def decrypt_raw(data: bytes, password: str) -> bytes:
    """Decrypt encrypted bytes in ATBE wire format.

    Args:
        data: ATBE wire bytes.
        password: Password used for PBKDF2 key derivation.

    Returns:
        Decrypted plaintext bytes.

    Raises:
        ATBDecryptionError: If the format is invalid or authentication fails.
    """
    if len(data) < HEADER_SIZE:
        raise ATBDecryptionError("invalid format")
    if data[: len(MAGIC)] != MAGIC:
        raise ATBDecryptionError("invalid format")
    version = data[len(MAGIC)]
    if not password:
        raise ATBDecryptionError("password cannot be empty")
    iterations = _kdf_iterations_for_version(version, error_cls=ATBDecryptionError)

    offset = len(MAGIC) + 1
    salt = data[offset : offset + SALT_SIZE]
    offset += SALT_SIZE
    nonce = data[offset : offset + NONCE_SIZE]
    offset += NONCE_SIZE
    tag = data[offset : offset + TAG_SIZE]
    offset += TAG_SIZE
    ciphertext = data[offset:]

    key = _derive_key(
        password,
        salt,
        iterations,
        error_cls=ATBDecryptionError,
    )
    aesgcm = AESGCM(key)
    try:
        return aesgcm.decrypt(nonce, ciphertext + tag, None)
    except InvalidTag as exc:
        raise ATBDecryptionError("authentication failed") from exc


def _bundle_payload(bundle: Bundle) -> dict[str, object]:
    records = [{"event": r.event, "hash": r.hash} for r in bundle.records]
    head_hash = records[-1]["hash"] if records else GENESIS_HASH
    return {"head_hash": head_hash, "records": records}


def encrypt_bundle(
    bundle: Bundle,
    password: str,
    *,
    salt: bytes | None = None,
    nonce: bytes | None = None,
) -> bytes:
    """Encrypt a bundle to ATBE bytes.

    Args:
        bundle: Bundle to verify and encrypt.
        password: Password used for PBKDF2 key derivation.
        salt: Optional deterministic salt for tests.
        nonce: Optional deterministic nonce for tests.

    Returns:
        ATBE wire bytes.

    Raises:
        ATBVerificationError: If the bundle fails hash-chain verification.
        ATBEncryptionError: If encryption fails.
    """
    bundle.verify()
    payload = _bundle_payload(bundle)
    canonical = canonicalize(payload)
    return encrypt_raw(canonical, password, salt=salt, nonce=nonce)


def decrypt_bundle(password: str, data: bytes) -> Bundle:
    """Decrypt bytes into a verified bundle.

    Args:
        password: Password used for PBKDF2 key derivation.
        data: ATBE wire bytes.

    Returns:
        A verified ``Bundle``.

    Raises:
        ATBDecryptionError: If decoding, authentication, or verification fails.
    """
    plaintext = decrypt_raw(data, password)
    try:
        payload = json.loads(plaintext.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ATBDecryptionError("decrypted payload is not valid JSON") from exc
    if not isinstance(payload, dict):
        raise ATBDecryptionError("decrypted payload must be an object")

    raw_records = payload.get("records")
    raw_head_hash = payload.get("head_hash")
    if not isinstance(raw_records, list) or not isinstance(raw_head_hash, str):
        raise ATBDecryptionError("decrypted payload missing required fields")

    from atb.bundle import Bundle, Record

    records: list[Record] = []
    for item in raw_records:
        if not isinstance(item, dict):
            raise ATBDecryptionError("record must be an object")
        event = item.get("event")
        record_hash = item.get("hash")
        if not isinstance(event, dict) or not isinstance(record_hash, str):
            raise ATBDecryptionError("record must include event object and hash string")
        records.append(Record(event=event, hash=record_hash))

    bundle = Bundle(records=records)
    try:
        bundle.verify()
    except ATBVerificationError as exc:
        raise ATBDecryptionError("decrypted payload failed hash-chain verification") from exc

    recomputed_head = bundle.records[-1].hash if bundle.records else GENESIS_HASH
    if raw_head_hash != recomputed_head:
        raise ATBDecryptionError(
            f"decrypted payload head_hash mismatch: expected {raw_head_hash}, got {recomputed_head}"
        )
    return bundle

File: sdk/typescript/src/encrypt.ts
import {
  createHash,
  createCipheriv,
  createDecipheriv,
  pbkdf2Sync,
  randomBytes,
} from "node:crypto";
import { canonicalize } from "./canonicalize.js";
import { parseEvent, prepareForCanonical, type Event } from "./event.js";
import { GENESIS_HASH } from "./hash.js";
import type { ATBRecord } from "./types.js";

export const MAGIC = "ATBE";
export const LEGACY_VERSION = 0x01;
export const VERSION = 0x02;
export const SALT_SIZE = 16;
export const NONCE_SIZE = 12;
export const TAG_SIZE = 16;
export const KEY_SIZE = 32;
export const LEGACY_PBKDF2_ITERATIONS = 100_000;
export const PBKDF2_ITERATIONS = 600_000;
export const HEADER_SIZE =
  MAGIC.length + 1 + SALT_SIZE + NONCE_SIZE + TAG_SIZE;

/** Error raised by encryption and decryption helpers. */
export class EncryptError extends Error {
  /**
   * @param message Human-readable encryption failure message.
   * @returns A new encryption error.
   */
  constructor(message: string) {
    super(message);
    this.name = "EncryptError";
  }
}

/** Optional deterministic encryption inputs for tests and golden vectors. */
export interface EncryptOptions {
  salt?: Uint8Array;
  nonce?: Uint8Array;
}

interface WirePayload {
  head_hash: string;
  records: ATBRecord[];
}

/** Decrypted and verified bundle payload. */
export interface DecryptedBundlePayload {
  headHash: string;
  records: ATBRecord[];
}

function deriveKey(
  password: string,
  salt: Uint8Array,
  iterations: number
): Buffer {
  if (!password) {
    throw new EncryptError("password cannot be empty");
  }
  if (salt.length !== SALT_SIZE) {
    throw new EncryptError(`salt must be ${SALT_SIZE} bytes`);
  }
  return pbkdf2Sync(password, salt, iterations, KEY_SIZE, "sha256");
}

function kdfIterationsForVersion(version: number): number {
  switch (version) {
    case LEGACY_VERSION:
      return LEGACY_PBKDF2_ITERATIONS;
    case VERSION:
      return PBKDF2_ITERATIONS;
    default:
      throw new EncryptError(
        `unsupported version: 0x${version.toString(16).padStart(2, "0")}`
      );
  }
}

function computeEventHash(event: Event, prevHash: string): string {
  return createHash("sha256")
    .update(prevHash, "utf8")
    .update(canonicalize(prepareForCanonical(event)), "utf8")
    .digest("hex");
}

/**
 * @param plaintext Plain bytes to encrypt.
 * @param password Password used for PBKDF2 key derivation.
 * @param options Optional deterministic salt and nonce.
 * @returns ATBE wire bytes.
 * @throws EncryptError when password, salt, nonce, or AES-GCM output is invalid.
 */
export function encryptRaw(
  plaintext: Uint8Array,
  password: string,
  options: EncryptOptions = {}
): Uint8Array {
  return encryptRawWithVersion(plaintext, password, VERSION, options);
}

function encryptRawWithVersion(
  plaintext: Uint8Array,
  password: string,
  version: number,
  options: EncryptOptions = {}
): Uint8Array {
  const salt = options.salt ? Buffer.from(options.salt) : randomBytes(SALT_SIZE);
  const nonce = options.nonce
    ? Buffer.from(options.nonce)
    : randomBytes(NONCE_SIZE);
  if (nonce.length !== NONCE_SIZE) {
    throw new EncryptError(`nonce must be ${NONCE_SIZE} bytes`);
  }

  const key = deriveKey(password, salt, kdfIterationsForVersion(version));
  const cipher = createCipheriv("aes-256-gcm", key, nonce);
  const ciphertext = Buffer.concat([
    cipher.update(Buffer.from(plaintext)),
    cipher.final(),
  ]);
  const tag = cipher.getAuthTag();
  if (tag.length !== TAG_SIZE) {
    throw new EncryptError(`auth tag must be ${TAG_SIZE} bytes`);
  }

  return Buffer.concat([
    Buffer.from(MAGIC, "ascii"),
    Buffer.from([version]),
    salt,
    nonce,
    tag,
    ciphertext,
  ]);
}

/**
 * @param data ATBE wire bytes.
 * @param password Password used for PBKDF2 key derivation.
 * @returns Decrypted plaintext bytes.
 * @throws EncryptError when format validation or authentication fails.
 */
export function decryptRaw(data: Uint8Array, password: string): Uint8Array {
  const bytes = Buffer.from(data);
  if (bytes.length < HEADER_SIZE) {
    throw new EncryptError("invalid format");
  }
  if (bytes.subarray(0, MAGIC.length).toString("ascii") !== MAGIC) {
    throw new EncryptError("invalid format");
  }
  const version = bytes[MAGIC.length];
  if (!password) {
    throw new EncryptError("password cannot be empty");
  }
  const iterations = kdfIterationsForVersion(version);

  let offset = MAGIC.length + 1;
  const salt = bytes.subarray(offset, offset + SALT_SIZE);
  offset += SALT_SIZE;
  const nonce = bytes.subarray(offset, offset + NONCE_SIZE);
  offset += NONCE_SIZE;
  const tag = bytes.subarray(offset, offset + TAG_SIZE);
  offset += TAG_SIZE;
  const ciphertext = bytes.subarray(offset);

  const key = deriveKey(password, salt, iterations);
  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAuthTag(tag);
  try {
    return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  } catch {
    throw new EncryptError("authentication failed");
  }
}

function toWirePayload(records: readonly ATBRecord[]): WirePayload {
  const headHash =
    records.length > 0
      ? records[records.length - 1].hash
      : GENESIS_HASH;
  return {
    head_hash: headHash,
    records: records.map((record) => ({
      event: parseEvent(record.event),
      hash: record.hash,
    })),
  };
}

function fromWirePayload(payload: unknown): DecryptedBundlePayload {
  if (!payload || typeof payload !== "object") {
    throw new EncryptError("decrypted payload must be an object");
  }
  const maybe = payload as Record<string, unknown>;
  const headHash = maybe.head_hash;
  const recordsValue = maybe.records;
  if (typeof headHash !== "string" || !Array.isArray(recordsValue)) {
    throw new EncryptError("decrypted payload missing required fields");
  }

  const parsedRecords: ATBRecord[] = recordsValue.map((item) => {
    if (!item || typeof item !== "object") {
      throw new EncryptError("record must be an object");
    }
    const record = item as Record<string, unknown>;
    if (typeof record.hash !== "string" || !record.event || typeof record.event !== "object") {
      throw new EncryptError("record must include event object and hash string");
    }
    let event: Event;
    try {
      event = parseEvent(record.event);
    } catch {
      throw new EncryptError("record event must include seq, prev_hash, and type");
    }
    return {
      event,
      hash: record.hash,
    };
  });

  const hasManifest =
    parsedRecords.length > 0 &&
    parsedRecords[0].event.type === "atb.bundle.manifest";
  let prev = GENESIS_HASH;
  for (let i = 0; i < parsedRecords.length; i++) {
    const record = parsedRecords[i];
    const expectedSeq = hasManifest ? (i === 0 ? 0 : i) : i + 1;
    if (record.event.seq !== expectedSeq || record.event.prev_hash !== prev) {
      throw new EncryptError("decrypted payload failed hash-chain verification");
    }
    const computed = computeEventHash(record.event, prev);
    if (computed !== record.hash) {
      throw new EncryptError("decrypted payload failed hash-chain verification");
    }
    prev = computed;
  }
  if (headHash !== prev) {
    throw new EncryptError(
      `decrypted payload head_hash mismatch: expected ${headHash}, got ${prev}`
    );
  }

  return {
    headHash,
    records: parsedRecords,
  };
}

/**
 * @param records Bundle records to verify and encrypt.
 * @param password Password used for PBKDF2 key derivation.
 * @param options Optional deterministic salt and nonce.
 * @returns Promise resolving to ATBE wire bytes.
 * @throws EncryptError when payload verification or encryption fails.
 */
export async function encryptBundle(
  records: readonly ATBRecord[],
  password: string,
  options: EncryptOptions = {}
): Promise<Uint8Array> {
  const payload = toWirePayload(records);
  const canonical = canonicalize(payload);
  return encryptRaw(Buffer.from(canonical, "utf8"), password, options);
}

/**
 * @param password Password used for PBKDF2 key derivation.
 * @param data ATBE wire bytes.
 * @returns Promise resolving to a verified decrypted payload.
 * @throws EncryptError when decryption, JSON decoding, or verification fails.
 */
export async function decryptBundle(
  password: string,
  data: Uint8Array
): Promise<DecryptedBundlePayload> {
  const plaintext = decryptRaw(data, password);
  let parsed: unknown;
  try {
    parsed = JSON.parse(Buffer.from(plaintext).toString("utf8")) as unknown;
  } catch {
    throw new EncryptError("decrypted payload is not valid JSON");
  }
  return fromWirePayload(parsed);
}
